### PR TITLE
DDF-4311 Remove UI component, add url functionality to the CLI, and u…

### PR DESCRIPTION
**Pull request by pvargas**
…pdate docs.

#### What does this PR do?
Removes the offline gazetteer from the UI, adds url functionality to the CLI, and updates the gazetteer documentation and  CLI help message to reflect the change.

*Edit:* @emmberk has added a fix for a related issue to this PR. The issue prevented loading geocode files.

#### Who is reviewing it? 
@mdang8 @kcover @peterhuffer @emmberk @oconnormi @stustison @troymohl 

#### Ask 2 committers to review/merge the PR and tag them here.

@andrewkfiedler
@clockard

#### How should this be tested?

- Try to add geocodes through the command line by using a link. 
`gazetteer:update linkToResource`  Links can be found [here](http://download.geonames.org/export/dump/)

- In the admin UI, verify the GeoNames Configuration tab under the Spatial app, is no longer there.

#### What are the relevant tickets?
[DDF-4311](https://codice.atlassian.net/browse/DDF-4311)

#### Screenshots
Before:
screen shot 2018-11-07 at 7 51 41 am

After:
screen shot 2018-11-07 at 8 29 27 am

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: distribution/docs/src/main/resources/content/_reference/_appReferences/mg-spatial.adoc [hunk 1]
```diff
@@ -27,7 +27,9 @@ The `gazetteer` commands provide the ability to interact with the local GeoNames
...
... GeoJSON data (GeoJSON URLs are not supported). Currently .txt and .zip files are supported for manual GeoNames entries and .geo.json files are supported for manual GeoJSON entries. GeoNames data can be manually downloaded from http://download.geonames.org/export/dump, where the absolute path of the file would be passed as an argument to the command (ex. /Users/johndoe/Downloads/AU.zip). Entries can also be automatically downloaded from http://download.geonames.org/export/dump by passing the country code as an argument to the command (ex. AU) which will add the country as GeoNames metacards in the core catalog. The full list of country codes available can be found in http://download.geonames.org/export/dump/countryInfo.txt. Using the argument "all" will download all of the current country codes (this process may take some time). In addition to country codes, GeoNames also provides entries for cities based on their population sizes. The arguments "cities1000", "cities5000", and "cities15000" will add cities to the index that have at least 1000, 5000, or 15000 people respectively. Additionally, URLs can also be used as an argument (ex. http://download.geonames.org/export/dump/AI.zip)
```
Comment (emmberk) [inline distribution/docs/src/main/resources/content/_reference/_appReferences/mg-spatial.adoc]: ✏️ `... people, respectively. Alternatively, a URL can be used as the argument (e.g. http://download.geonames.org/export/dump/AI.zip).`

File: catalog/spatial/spatial-app/src/main/resources/features.xml [hunk 2]
```diff
@@ -145,7 +145,6 @@
         <bundle>mvn:org.codice.ddf.spatial/spatial-geocoding-offline-catalog/${project.version}</bundle>
         <bundle>mvn:org.codice.ddf.spatial/spatial-geocoding-feature/${project.version}</bundle>
         <bundle>mvn:org.codice.ddf.spatial/spatial-commands/${project.version}</bundle>
-        <bundle>mvn:org.codice.ddf.spatial/spatial-geocoding-admin-module/${project.version}</bundle>
```
Comment (emmberk) [inline catalog/spatial/spatial-app/src/main/resources/features.xml]: Should _catalog/spatial/geocoding/spatial-geocoding-admin-module_ and _ui/packages/spatial-geocoding-admin-module_ be deleted too?

What is the motivation for removing the Admin Console part instead of fixing it?

Comment (pvargas) [reply catalog/spatial/spatial-app/src/main/resources/features.xml]: One of the endpoints used by the UI no longer exists. It was removed a long time ago and apparently no one noticed, so @clockard said to just remove the UI part.

Comment (clockard) [reply catalog/spatial/spatial-app/src/main/resources/features.xml]: We should remove the modules/code as well. We don't want to leave dead code hanging around.

File: catalog/spatial/geowebcache/README.md [hunk 3]
```diff
@@ -59,7 +59,7 @@ To update a configuration with CURL :
 GeoWebCache also provides a user interface that can be used to preview layers and truncate or seed them at `https://localhost:8993/geowebcache/`.
 
 ### Configure GWC disk quota
-> Since disk usage increases geometrically by zoom level, one single seeding task could fill up an entire storage device. Because of this, GeoWebCache employs a disk quota system where one can specify the maximum amount of disk space to use for a particular layer or for the entire set of layers (the “Global Quota”), as well as logic on how to proceed when that quota is reached. There are two different policies for managing the disk quotas: Least Frequently Used (LFU) and Least Recently Used (LRU). 
+> Since disk usage increases geometrically by zoom level, one single seeding task could fill up an entire storage device. Because of this, GeoWebCache employs a disk quota system where one can specify the maximum amount of disk space to use for a particular layer or for the entire set of layers (the “Global Quota”), as well as logic on how to proceed when that quota is reached. There are two different policies for managing the disk quotas: Least Frequently Used (LFU) and Least Recently Used (LRU).
```
Comment (clockard) [inline catalog/spatial/geowebcache/README.md]: Is the only change here a space at the end of the line? If so can we revert this change since it isn't related to the rest of the change set.

File: catalog/spatial/geocoding/spatial-geocoding-offline-catalog/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesCatalogIndexer.java [hunk 4]
```diff
@@ -142,15 +142,6 @@ private Metacard transformGeoEntryToMetacard(GeoEntry geoEntry) {
     metacard.setAttribute(
         new AttributeImpl(
             GeoEntryAttributes.FEATURE_CODE_ATTRIBUTE_NAME, geoEntry.getFeatureCode()));
-    Integer gazetteerSortValue = getGeoNameGazetterSortByFeatureClass(geoEntry);
```
Comment (clockard) [inline catalog/spatial/geocoding/spatial-geocoding-offline-catalog/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesCatalogIndexer.java]: Why was this removed? Not sure what it is trying to accomplish but seems unrelated to the overall goal of the pr.

File: catalog/spatial/geowebcache/README.md (hunk 3 above)
Comment (pvargas) [reply catalog/spatial/geowebcache/README.md]: Yeah, this was the formatter.

File: catalog/spatial/geocoding/spatial-geocoding-offline-catalog/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesCatalogIndexer.java (hunk 4 above)
Comment (pvargas) [reply catalog/spatial/geocoding/spatial-geocoding-offline-catalog/src/main/java/org/codice/ddf/spatial/geocoding/index/GeoNamesCatalogIndexer.java]: It was from a merge conflict. I'll revert that.

File: catalog/spatial/geocoding/spatial-geocoding-extract/pom.xml [hunk 5]
```diff
@@ -73,6 +92,10 @@
                     <instructions>
                         <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
                         <Embed-Dependency>
+                            platform-utils,
+                            commons-validator,
```
Comment (peterhuffer) [inline catalog/spatial/geocoding/spatial-geocoding-extract/pom.xml]: ✏️ We shouldn't be embedding commons-validator, commons-beanutils or commons-digester. They are bundles. Instead we should make sure it is started with this feature.

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 6]
```diff
@@ -248,10 +257,14 @@ private InputStream getInputStreamFromUrl(
   }
 
   public Response createConnection(String resource) {
-    webClient = WebClient.create(url + resource + ".zip");
-    webClient.path(url + resource + ".zip");
-    Response response = webClient.get();
-    return response;
+    if (urlValidator.isValid(resource)) {
+      webClient = WebClient.create(resource);
```
Comment (peterhuffer) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓ Should we be using a Secure CXF client here? @stustison

File: catalog/spatial/geocoding/spatial-geocoding-offline-catalog/pom.xml [hunk 7]
```diff
@@ -176,7 +191,11 @@
                             spatial-geocoding-api-impl,
                             spatial4j,
                             noggit,
-                            platform-util
+                            platform-util,
+                            platform-utils,
+                            commons-validator,
```
Comment (peterhuffer) [inline catalog/spatial/geocoding/spatial-geocoding-offline-catalog/pom.xml]: ✏️ Same comment about embedding these commons libraries. Also, I don't see `platform-utils`, does that exist?

File: catalog/spatial/geocoding/spatial-geocoding-offline-catalog/pom.xml [hunk 8]
```diff
@@ -199,7 +218,7 @@
                             <rules>
                                 <ArtifactSizeEnforcerRule
                                     implementation="org.codice.maven.artifactsize.ArtifactSizeEnforcerRule">
-                                    <maxArtifactSize>6.7_MB</maxArtifactSize>
+                                    <maxArtifactSize>6.9_MB</maxArtifactSize>
```
Comment (peterhuffer) [inline catalog/spatial/geocoding/spatial-geocoding-offline-catalog/pom.xml]: ✏️ We can probably lower this back down once we don't embed those dependencies.

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 9]
```diff
@@ -144,15 +148,20 @@ public void pushGeoEntriesToExtractionCallback(
...
+    } else if (!urlValidator.isValid(resource) && FilenameUtils.isExtension(resource, "txt")) {
       return getFileInputStream(resource);
 
-    } else if (StringUtils.isAlphanumeric(resource)) {
+    } else if (StringUtils.isNotBlank(resource)) {
+
+      if (!StringUtils.isAlphanumeric(resource) && !urlValidator.isValid(resource)) {
+        throw new GeoEntryExtractionException(
+            String.format(
+                "Unable to update the index. The file at %s is not a supported path.", resource));
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ✏️ `Unable to update the index. %s is not a supported file path.`

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 10]
```diff
@@ -144,15 +148,20 @@ public void pushGeoEntriesToExtractionCallback(
   private InputStream getInputStreamFromResource(
       String resource, ProgressCallback extractionCallback)
       throws GeoEntryExtractionException, GeoNamesRemoteDownloadException {
-
-    if (FilenameUtils.isExtension(resource, "zip")) {
+    if (!urlValidator.isValid(resource) && FilenameUtils.isExtension(resource, "zip")) {
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓ Why do we need to check urlValidator.isValid(resource) in this method? Can you extract it to a variable so that you don't need to check it multiple times?

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 11]
```diff
@@ -161,7 +170,7 @@ private InputStream getInputStreamFromResource(
       } else if (resource.matches("((?i)cities[0-9]+)")) {
         resource = resource.toLowerCase();
         // Support case insensitive country codes
-      } else if (!resource.equalsIgnoreCase("allCountries")) {
+      } else if (!resource.equalsIgnoreCase("allCountries") && !urlValidator.isValid(resource)) {
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓ Do you need to check if it is a valid URL here if you already know that the string is "allCountries"?

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 12]
```diff
@@ -248,10 +257,14 @@ private InputStream getInputStreamFromUrl(
   }
 
   public Response createConnection(String resource) {
-    webClient = WebClient.create(url + resource + ".zip");
-    webClient.path(url + resource + ".zip");
-    Response response = webClient.get();
-    return response;
+    if (urlValidator.isValid(resource)) {
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓ It looks like this method is mocked in the unit tests and never gets tested. Does the urlValidator need to be tested here?

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java (hunk 10 above)
Comment (emmberk) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓ Are file paths URLs? What if someone entered file://...? Would this be handled by `webClient = WebClient.create(resource);`?

File: catalog/spatial/spatial-commands/src/main/java/org/codice/ddf/commands/spatial/gazetteer/GazetteerUpdateCommand.java [hunk 13]
```diff
@@ -58,7 +58,8 @@
             + "When the resource is a path to a file, it will be imported locally."
             + "If a path to a file ends in .geo.json, it will processed as a"
             + "geoJSON feature collection and imported as supplementary shape "
-            + "data for geonames entries.",
+            + "data for geonames entries. Additionally, the resource can also be a URL "
+            + "(ex: http://download.geonames.org/export/dump/AI.zip)",
```
Comment (emmberk) [inline catalog/spatial/spatial-commands/src/main/java/org/codice/ddf/commands/spatial/gazetteer/GazetteerUpdateCommand.java]: ✏️ Missing a period at the end of this sentence. This description should match the description in the documentation.

File: catalog/spatial/spatial-app/src/main/resources/features.xml (hunk 2 above)
Comment (emmberk) [reply catalog/spatial/spatial-app/src/main/resources/features.xml]: Should catalog/spatial/geocoding/spatial-geocoding-admin-module be deleted too?

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 14]
```diff
@@ -51,6 +52,9 @@
 
   private String url;
 
+  private String[] schemes = {"http", "https", "ALLOW_ALL_SCHEMES"};
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓How does this `ALLOW_ALL_SCHEMES` work? Should it be an option like `UrlValidator.ALLOW_LOCAL_URLS`? Why do you specify `ALLOW_ALL_SCHEMES` and also http/https?

File: catalog/spatial/spatial-app/src/main/resources/features.xml (hunk 2 above)
Comment (pvargas) [reply catalog/spatial/spatial-app/src/main/resources/features.xml]: @emmberk You're right! I thought that was the config, but that is elsewhere.

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java (hunk 10 above)
Comment (emmberk) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: 🤷‍♀️ Maybe it'd be worth adding a new unit test for this case.

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java (hunk 12 above)
Comment (oconnormi) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: Yeah there are two methods that are overridden by the `Spy` from what I can see. Both `getUrlInputStreamFromWebClient` and `createConnection` are never touched by any of the exisiting tests. This class would need to be refactored a bit to make it easier to test it. I think one of the main problems I noticed is that there is no way to provide a mock `WebClient` since the `createConnection` method makes a call to `WebClient.create`. It looks like @pvargas followed the pattern of the existing tests here. Should he refactor this class a bit to make it possible to test everything?

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 15]
```diff
@@ -161,7 +174,7 @@ private InputStream getInputStreamFromResource(
       } else if (resource.matches("((?i)cities[0-9]+)")) {
         resource = resource.toLowerCase();
         // Support case insensitive country codes
-      } else if (!resource.equalsIgnoreCase("allCountries")) {
+      } else if (!resource.equalsIgnoreCase("allCountries") && !resourceIsValidUrl) {
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ✏️ Is this `&& !resourceIsValidUrl` needed here?

File: catalog/spatial/spatial-app/src/main/resources/features.xml (hunk 2 above)
Comment (emmberk) [reply catalog/spatial/spatial-app/src/main/resources/features.xml]: ✏️ Does this line need to be removed too? https://github.com/codice/ddf/blob/851f92fccfad58b54f960aa8b1b0c1bd773d1350/catalog/spatial/geocoding/pom.xml#L36

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java (hunk 12 above)
Comment (emmberk) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: I think manual testing might be sufficient in this case.

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java (hunk 15 above)
Comment (pvargas) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: Yes, otherwise valid URLs become uppercase. I had to work with the existing logic.

Comment (emmberk) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: Isn't it a guarantee that all uppercase/lowercase variations of the String "allCountries" are not valid URLs?

Comment (pvargas) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: I think the condition is evaluating if the string is not "allCountries" (regardless of capitalization). If the string is not "allCountries" it is not guaranteed be a valid URL.

Comment (emmberk) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: 🤦‍♀️ You're right. I misread that.

**Comment by pvargas** [comment]
Build now!

**Comment by pvargas** [comment]
Build now, please.

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 16]
```diff
@@ -41,8 +44,14 @@
 import org.codice.ddf.spatial.geocoding.ProgressCallback;
 
 public class GeoNamesFileExtractor implements GeoEntryExtractor {
+
   private static final int BUFFER_SIZE = 4096;
 
+  private static final String[] SCHEMES = {"http", "https"};
+
+  private static final UrlValidator URL_VALIDATOR =
+      new UrlValidator(SCHEMES, UrlValidator.ALLOW_LOCAL_URLS);
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓Is this UrlValidator really needed? Why not just try the resource as a file path, and , if that fails, try the resource as a URl, and, if that fails, try the resource as a keyword to automatically download a zip file from http://download.geonames.org/export/dump? You wouldn't need to start those extra three bundles.

File: catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java [hunk 17]
```diff
@@ -161,7 +176,7 @@ private InputStream getInputStreamFromResource(
       } else if (resource.matches("((?i)cities[0-9]+)")) {
         resource = resource.toLowerCase();
```
Comment (emmberk) [inline catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: ❓I know that these aren't your changes, but why do we need to modify the case here for country codes and cities? Doesn't it just make a URL connection and URLs are case-insensitive?

Comment (pvargas) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: The filename part of the URL is case sensitive.

Comment (emmberk) [reply catalog/spatial/geocoding/spatial-geocoding-extract/src/main/java/org/codice/ddf/spatial/geocoding/extract/GeoNamesFileExtractor.java]: Oh, good point. I just confirmed that `http://download.geonames.org/export/dump/aU.zip` doesn't work.

**Comment by pvargas** [comment]
Build now ( ͡° ͜ʖ ͡°)
